File: app/src/main/java/com/iita/akilimo/utils/enums/EnumPotatoUnitPrice.java
package com.iita.akilimo.utils.enums;


import com.iita.akilimo.utils.MathHelper;

public enum EnumPotatoUnitPrice {
    UNKNOWN {
        double unitPriceLower = -1;
        double unitPriceUpper = -1;

        @Override
        public double convertToLocalCurrency(String toCurrency, MathHelper mathHelper) {
            return -1;
        }

        @Override
        public double unitPricePerTonneLower() {
            return unitPriceLower;
        }

        @Override
        public double unitPricePerTonneUpper() {
            return unitPriceUpper;
        }
    },
    PRICE_RANGE_ONE {
        double unitPriceLower = 47.62;
        double unitPriceUpper = 49.79;

        @Override
        public double convertToLocalCurrency(String toCurrency, MathHelper mathHelper) {
            double usd = (unitPriceLower + unitPriceUpper) / 2;
            return EnumPotatoUnitPrice.convertCurrency(usd, toCurrency, mathHelper);
        }

        @Override
        public double unitPricePerTonneLower() {
            return unitPriceLower;
        }

        @Override
        public double unitPricePerTonneUpper() {
            return unitPriceUpper;
        }
    },
    PRICE_RANGE_TWO {
        double unitPriceLower = 49.79;
        double unitPriceUpper = 51.95;

        @Override
        public double convertToLocalCurrency(String toCurrency, MathHelper mathHelper) {
            double usd = (unitPriceLower + unitPriceUpper) / 2;
            return EnumPotatoUnitPrice.convertCurrency(usd, toCurrency, mathHelper);
        }

        @Override
        public double unitPricePerTonneLower() {
            return unitPriceLower;
        }

        @Override
        public double unitPricePerTonneUpper() {
            return unitPriceUpper;
        }
    },
    PRICE_RANGE_THREE {
        double unitPriceLower = 51.95;
        double unitPriceUpper = 56.28;

        @Override
        public double convertToLocalCurrency(String toCurrency, MathHelper mathHelper) {
            double usd = (unitPriceLower + unitPriceUpper) / 2;
            return EnumPotatoUnitPrice.convertCurrency(usd, toCurrency, mathHelper);
        }

        @Override
        public double unitPricePerTonneLower() {
            return unitPriceLower;
        }

        @Override
        public double unitPricePerTonneUpper() {
            return unitPriceUpper;
        }
    },
    PRICE_EXACT {
        @Override
        public double convertToLocalCurrency(String toCurrency, MathHelper mathHelper) {
            return 0.0;
        }

        @Override
        public double unitPricePerTonneLower() {
            return 0;
        }

        @Override
        public double unitPricePerTonneUpper() {
            return 0;
        }
    };

    public abstract double convertToLocalCurrency(String toCurrency, MathHelper mathHelper);

    public abstract double unitPricePerTonneLower();

    public abstract double unitPricePerTonneUpper();

    private static double convertCurrency(double amount, String toCurrency, MathHelper mathHelper) {
        return mathHelper.convertToLocalCurrency(amount, toCurrency);
    }
}

File: app/src/main/java/com/iita/akilimo/utils/enums/EnumPotatoUnitPrice.kt
package com.iita.akilimo.utils.enums

import com.iita.akilimo.utils.MathHelper

enum class EnumPotatoUnitPrice {
    UNKNOWN {
        var unitPriceLower = -1.0
        var unitPriceUpper = -1.0
        override fun convertToLocalCurrency(
            toCurrency: String?,
            mathHelper: MathHelper?
        ): Double {
            return (-1).toDouble()
        }

        override fun unitPricePerTonneLower(): Double {
            return unitPriceLower
        }

        override fun unitPricePerTonneUpper(): Double {
            return unitPriceUpper
        }
    },
    PRICE_RANGE_ONE {
        var unitPriceLower = 47.62
        var unitPriceUpper = 49.79
        override fun convertToLocalCurrency(
            toCurrency: String?,
            mathHelper: MathHelper?
        ): Double {
            val usd = (unitPriceLower + unitPriceUpper) / 2
            return convertCurrency(usd, toCurrency, mathHelper)
        }

        override fun unitPricePerTonneLower(): Double {
            return unitPriceLower
        }

        override fun unitPricePerTonneUpper(): Double {
            return unitPriceUpper
        }
    },
    PRICE_RANGE_TWO {
        var unitPriceLower = 49.79
        var unitPriceUpper = 51.95
        override fun convertToLocalCurrency(
            toCurrency: String?,
            mathHelper: MathHelper?
        ): Double {
            val usd = (unitPriceLower + unitPriceUpper) / 2
            return convertCurrency(usd, toCurrency, mathHelper)
        }

        override fun unitPricePerTonneLower(): Double {
            return unitPriceLower
        }

        override fun unitPricePerTonneUpper(): Double {
            return unitPriceUpper
        }
    },
    PRICE_RANGE_THREE {
        var unitPriceLower = 51.95
        var unitPriceUpper = 56.28
        override fun convertToLocalCurrency(
            toCurrency: String?,
            mathHelper: MathHelper?
        ): Double {
            val usd = (unitPriceLower + unitPriceUpper) / 2
            return convertCurrency(usd, toCurrency, mathHelper)
        }

        override fun unitPricePerTonneLower(): Double {
            return unitPriceLower
        }

        override fun unitPricePerTonneUpper(): Double {
            return unitPriceUpper
        }
    },
    PRICE_EXACT {
        override fun convertToLocalCurrency(
            toCurrency: String?,
            mathHelper: MathHelper?
        ): Double {
            return 0.0
        }

        override fun unitPricePerTonneLower(): Double {
            return 0.0
        }

        override fun unitPricePerTonneUpper(): Double {
            return 0.0
        }
    };

    abstract fun convertToLocalCurrency(
        toCurrency: String?,
        mathHelper: MathHelper?
    ): Double

    abstract fun unitPricePerTonneLower(): Double
    abstract fun unitPricePerTonneUpper(): Double

    companion object {
        private fun convertCurrency(
            amount: Double,
            toCurrency: String?,
            mathHelper: MathHelper?
        ): Double {
            val resp = when (mathHelper) {
                null -> {
                    val mh = MathHelper()
                    mh.convertToLocalCurrency(amount, toCurrency)
                }
                else -> {
                    mathHelper.convertToLocalCurrency(amount, toCurrency)
                }
            }

            return resp
        }
    }
}